File: solutions/typescript/resistor-color-trio/1/resistor-color-trio.ts
const COLORS = "black brown red orange yellow green blue violet grey white".split(" ")

export function decodedResistorValue(colors: string[]) : string {
  const [first, second, third] = colors
  return encode(decodeTwo(first, second) * calcMagnitude(third))
}

function decodeTwo(first:string, second:string) : number {
  return COLORS.indexOf(first) * 10 + COLORS.indexOf(second)
}

function calcMagnitude(third:string) : number {
  return 10 ** COLORS.indexOf(third)
}

function encode(total:number) : string {
  if ( total >= 1000) {
    return `${total/1000} kiloohms`
  }
  return `${total} ohms`
}

File: solutions/elixir/resistor-color-trio/1/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do
  @colors ~w(black brown red orange yellow green blue violet grey white)a
  @doc """
  Calculate the resistance value in ohm or kiloohm from resistor colors
  """
  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms}
  def label(colors) do
    digits =
      colors
      |> Enum.take(3)
      |> Enum.map(fn color -> Enum.find_index(@colors, &(&1 == color)) end)

    [first, second | rest] = digits
    base = [first, second] |> Enum.join() |> String.to_integer()
    ret = base * 10 ** List.first(rest)

    if rem(ret, 1000) == 0 do
      {div(ret, 1000), :kiloohms}
    else
      {ret, :ohms}
    end
  end
end
